airflow/models/renderedtifields.py

```diff
@@ -0,0 +1,172 @@
+        """Checks templated field exist for this ti.
+
+        :param ti: Task Instance
+        :param session: SqlAlchemy Session
+        """
+        # We cant use q.exists() because some databases such as SQL Server don’t allow
+        # an EXISTS expression to be present in the columns clause of a SELECT.
+        # Details: https://docs.sqlalchemy.org/en/13/orm/query.html#sqlalchemy.orm.query.Query.exists
+        return session.query(exists().where(
+            and_(cls.dag_id == ti.dag_id,
+                 cls.task_id == ti.task_id,
+                 cls.execution_date == ti.execution_date)
+        )).scalar()
+
+    @provide_session
+    def write(self, overwrite: bool = False, session: Session = None):
+        """Write instance to database
+
+        :param overwrite: Override if record exists in DB
+        :param session: SqlAlchemy Session
+        """
+        if not self.has_templated_fields(self.ti, session):
+            session.add(self)
+        else:
+            if overwrite:
+                result = (
+                    session
+                    .query(RenderedTaskInstanceFields)
+                    .filter(
+                        RenderedTaskInstanceFields.dag_id == self.dag_id,
+                        RenderedTaskInstanceFields.task_id == self.task_id,
+                        RenderedTaskInstanceFields.execution_date == self.execution_date
+                    )
+                    .with_for_update(of=RenderedTaskInstanceFields)
+                    .first()
+                )
+
+                # Overwrite rendered_fields column
+                result.rendered_fields = self.rendered_fields
+                session.merge(self)
+
+    @classmethod
+    @provide_session
+    def delete_old_records(
+        cls, task_id: str, dag_id: str,
+        num_to_keep=conf.getint("core", "max_num_rendered_ti_fields_per_task", fallback=0),
+        session: Session = None
+    ):
+        """
+        Keep only Last X (num_to_keep) number of records for a task by deleting others
+
+        :param task_id: Task ID
+        :param dag_id: Dag ID
+        :param num_to_keep: Number of Records to keep
+        :param session: SqlAlchemy Session
+        """
+        if num_to_keep > 0:
+            # Fetch Top X records given dag_id & task_id ordered by Execution Date
+            subq1 = (
+                session
+                .query(cls.dag_id, cls.task_id, cls.execution_date)
+                .filter(cls.dag_id == dag_id, cls.task_id == task_id)
+                .order_by(cls.execution_date.desc())
+                .limit(num_to_keep)
+                .subquery('subq1')
+            )
+
+            # Second Subquery
+            # Workaround for MySQL Limitation (https://stackoverflow.com/a/19344141/5691525)
+            # Limitation: This version of MySQL does not yet support
+            # LIMIT & IN/ALL/ANY/SOME subquery
+            subq2 = (
```

Comment:
Is it worth only doing this for mysql?

Reply:
Not worth adding additional check to see if it is just SQL, it would be server side query, so not much gains.

The only extra bit is **subq2** which is just basically selecting everything from **subq1** so I don't think it would be expensive. WDYT